internal/impl/observability.go

```diff
@@ -0,0 +1,1449 @@
+	//
+	// [1] https://grafana.com/docs/loki/latest/clients/promtail/
+	promtailImageName = "grafana/promtail"
+
+	// Grafana related configs.
+
+	// The name of the Grafana [1] image used to display metrics, traces, and logs.
+	//
+	// [1] https://grafana.com/
+	grafanaImageName = "grafana/grafana"
+
+	// The default Grafana web server port.
+	grafanaPort = 3000
+)
+
+// dashboard was generated using the Grafana UI. Then, we saved the content as
+// a JSON file.
+//
+//go:embed dashboard.txt
+var dashboardContent string
+
+// generateObservabilityInfo generates deployment information needed by the app
+// to export metrics, logs, and traces.
+func generateObservabilityInfo(dep *protos.Deployment, cfg *KubeConfig) ([]byte, error) {
+	content, err := generateInfoToExportTraces(dep, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create kube info to export traces: %w", err)
+	}
+	var generated []byte
+	generated = append(generated, content...)
+
+	content, err = generateInfoToExportMetrics(dep, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create kube info to export metrics: %w", err)
+	}
+	generated = append(generated, content...)
+
+	content, err = generateInfoToExportLogs(dep, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create kube info to export logs: %w", err)
+	}
+	generated = append(generated, content...)
+
+	// Generate deployment info to export logs, traces and metrics to Grafana.
+	content, err = generateInfoToExportToGrafana(dep, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create kube info to export data to Grafana: %w", err)
+	}
+	generated = append(generated, content...)
+	return generated, nil
+}
+
+// generateInfoToExportTraces generates the Jaeger kubernetes deployment
+// information for a given app.
+//
+// Note that we run a single instance of Jaeger. This is because we are using
+// a Jaeger image that combines three Jaeger components, agent, collector, and
+// query service/UI in a single image.
+//
+// TODO(rgrandl): If the trace volume can't be handled by a single instance, we
+// should scale these components independently, and use different image(s).
+//
+// TODO(rgrandl): Convert the below comments into docs.
+// How to integrate with an external Jaeger service?
+// E.g., if you use Helm [1] to install Jaeger, you can simply do the following:
+// 1) You install Jaeger using a command similar to the one below.
+// helm install jaeger-all-in-one jaeger-all-in-one/jaeger-all-in-one
+//
+// 2) Your Jaeger service has the name 'jaeger-all-in-one'.
+//
+// 3) in your config.toml, set the 'jaeger_service' info as follows:
+// config.toml
+// ...
+// [kube]
+// observability = {jaeger_service = "jaeger-all-in-one"}
+//
+// [1] https://helm.sh/
+func generateInfoToExportTraces(dep *protos.Deployment, cfg *KubeConfig) ([]byte, error) {
+	// The user disabled exporting the traces, don't generate anything.
+	if !shouldGenerateKubeDeploymentInfo(exportTracesURL, cfg) {
```

Comment:
This confused the life out of me, since `exportTracesURL` sounds like it represents an actual URL to the Jaeger endpont. Instead, it's the key in the `Observability` map.

Can you rename:

`s/exportTracesURL/tracesConfigKey/'

Same for metrics etc.

Actually, even though it's somewhat prone to error, I would be happier to just inline "trace_service" here since it would be clearer.